consensus/cosipbft/forward_link.go

```diff
@@ -0,0 +1,262 @@
+type Digest = []byte
+
+// forwardLink is the cryptographic primitive to ensure a block is a successor
+// of a previous one.
+type forwardLink struct {
+	hash Digest
+	from Digest
+	to   Digest
+	// prepare signs the combination of From and To to prove that the nodes
+	// agreed on a valid forward link between the two blocks.
+	prepare crypto.Signature
+	// commit signs the Prepare signature to prove that a threshold of the
+	// nodes have committed the block.
+	commit crypto.Signature
+}
+
+// Verify makes sure the signatures of the forward link are correct.
+func (fl forwardLink) Verify(v crypto.Verifier, pubkeys []crypto.PublicKey) error {
+	err := v.Verify(pubkeys, fl.hash, fl.prepare)
+	if err != nil {
+		return xerrors.Errorf("couldn't verify prepare signature: %w", err)
+	}
+
+	buffer, err := fl.prepare.MarshalBinary()
+	if err != nil {
+		return xerrors.Errorf("couldn't marshal the signature: %w", err)
+	}
+
+	fabric.Logger.Trace().Msgf("verifying commit %x", buffer)
+	err = v.Verify(pubkeys, buffer, fl.commit)
+	if err != nil {
+		return xerrors.Errorf("couldn't verify commit signature: %w", err)
+	}
+
+	return nil
+}
+
+func encodeSignature(sig crypto.Signature) (*any.Any, error) {
+	packed, err := sig.Pack()
+	if err != nil {
+		return nil, xerrors.Errorf("couldn't pack: %v", err)
+	}
+
+	packedAny, err := protoenc.MarshalAny(packed)
+	if err != nil {
+		return nil, encoding.NewAnyEncodingError(packed, err)
+	}
+
+	return packedAny, nil
+}
+
+// Pack returns the protobuf message of the forward link.
+func (fl forwardLink) Pack() (proto.Message, error) {
+	seal := &ForwardLinkProto{
+		From: fl.from,
+		To:   fl.to,
+	}
+
+	var err error
+
+	if fl.prepare != nil {
+		seal.Prepare, err = encodeSignature(fl.prepare)
+		if err != nil {
+			return nil, encoding.NewEncodingError("prepare", err)
+		}
+	}
+
+	if fl.commit != nil {
+		seal.Commit, err = encodeSignature(fl.commit)
+		if err != nil {
+			return nil, encoding.NewEncodingError("commit", err)
+		}
+	}
+
+	return seal, nil
+}
+
+func (fl forwardLink) computeHash() ([]byte, error) {
+	// TODO: optional algorithm + errors
+	h := sha256.New()
+
+	h.Write(fl.from)
+	h.Write(fl.to)
+
+	return h.Sum(nil), nil
+}
+
+type forwardLinkChain struct {
+	links []forwardLink
+}
+
+// Verify follows the chain from the beginning and makes sure that the forward
+// links are correct and that they point to the correct targets.
+func (c forwardLinkChain) Verify(verifier crypto.Verifier, pubkeys []crypto.PublicKey) error {
+	if len(c.links) == 0 {
+		return xerrors.New("chain is empty")
+	}
+
+	lastIndex := len(c.links) - 1
+
+	for i, link := range c.links {
+		err := link.Verify(verifier, pubkeys)
+		if err != nil {
+			return xerrors.Errorf("couldn't verify link %d: %w", i, err)
+		}
+
+		if i != lastIndex && !bytes.Equal(link.to, c.links[i+1].from) {
+			return xerrors.Errorf("mismatch forward link '%x' != '%x'",
+				link.to, c.links[i+1].from)
+		}
+	}
```

Comment:
This is doing to many array calls (although it might be optimized by the compiler). What if I rename link as `currentLink` and add a line `nextLink := c.links[i+1]` ?